picard/tagger.py

```diff
@@ -511,81 +547,118 @@ def add_files(self, filenames, target=None):
-                self.unclustered_files.add_files(new_files)
+            if target is self.unclustered_files:
                 target = None
+
             for file in new_files:
-                file.load(partial(self._file_loaded, target=target))
+                self._pending_files[file.filename] = (file, target)
+                self._pending_files_process_queue.put(file.filename)
 
-    def add_directory(self, path):
-        if config.setting['recursively_add_files']:
-            self._add_directory_recursive(path)
-        else:
-            self._add_directory_non_recursive(path)
+    def _file_load_finished_thread(self):
+        while True:
+            filename, metadata = self._pending_files_results_queue.get()
 
-    def _add_directory_recursive(self, path):
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        walk = os.walk(path)
+            if self.stopping:
+                break
+
+            if filename in self._pending_files:
+                file, target = self._pending_files.pop(filename)
+
+                error = not metadata or None
+                thread.to_main(file._loading_finished, result=metadata, error=error, callback=None)
+
+                time.sleep(0)  # Yield from the CPU quantum and let other threads work
+                self._file_loaded_queue.put((file, target))
+
+        # Append empty message to kill _file_loaded_thread
+        self._file_loaded_queue.put((None, None))
+
+    def _file_loaded_thread(self):
+        while True:
+            file, target = self._file_loaded_queue.get()
+
+            if self.stopping:
+                break
+
+            thread.to_main(self._file_loaded, file, target)
+            time.sleep(0.01)  # Yields from CPU quantum, should be enough time for clustering few files without freezing
+
+    def _loader_process_watchdog_thread(self):
+        exitcode = None
+        while not self.stopping:
+            # Check if worker process died
+            if hasattr(self._loader_process, "exitcode") and self._loader_process.exitcode is not None:
+                exitcode = self._loader_process.exitcode
+                log.warning("Worker process died with exit code '%d'" %
+                             exitcode)
+
+            # (Re)Start worker
+            if self._loader_process is None or exitcode:
+                self._loader_process = mp.Process(target=File._load_check_metadata,
+                                                  args=(self._process_stopping,
+                                                        self._pending_files_process_queue,
+                                                        self._pending_files_results_queue))
+                self._loader_process.start()
+
+            # In case the process died previously, queue will be empty, so reinclude pending files
+            if exitcode is not None:
+                exitcode = None
+                for filename in self._pending_files.keys():
```

Comment:
Since `self._pending_files` is a dict, you can just write `for filename in self._pending_files:` (it iterates over keys by default).

Reply:
OK